Add dynamic registration for Windows desktops docs (#48968)
gravitational/teleport: docs/pages/enroll-resources/desktop-access/dynamic-registration.mdx

---
title: Dynamic Windows Desktop Registration
description: Register/unregister Windows desktops without restarting Teleport.
---

Dynamic Windows desktop registration allows Teleport administrators to register new Windows desktops (or
update/unregister existing ones) without having to update the static
configuration files read by Teleport Windows Desktop Service instances.

> Windows Desktop Service instances periodically query the Teleport Auth Service for
> `dynamic_windows_desktop` resources, each of which includes the information that the Windows Desktop

Reviewer: ```suggestion
Windows Desktop Service instances watch for updates from the Teleport Auth Service for
`dynamic_windows_desktop` resources, each of which includes the information that the Windows Desktop
```

Service needs to connect to a Windows desktop.

## Required permissions

In order to interact with dynamically registered Windows desktops, a user must have

> a Teleport role with permissions to manage `dynamic_windows_desktop` resources.

Reviewer: Do any of the preset roles have this permission by default, or do we always have to add it?

Is this behavior the same as for apps/databases?

```yaml
windows_desktop_service:
  enabled: "yes"
  resources:
  - labels:
      "*": "*"
```

You can use a wildcard selector to register all dynamic Windows desktop resources in the cluster
on the Windows Desktop Service or provide a specific set of labels for a subset:

```yaml
resources:
- labels:
    "env": "prod"
- labels:
    "env": "test"
```

## Creating a dynamic_windows_desktop resource

Configure Teleport to register a Windows desktop dynamically by creating an `dynamic_windows_desktop`
resource. The following example configures Teleport for connecting to Windows desktop
called `example` at `localhost:3089`.

```yaml
kind: dynamic_windows_desktop
version: v1
metadata:
  name: example
  description: "Example desktop"
  labels:
    env: test
spec:
  addr: localhost:3089
```

The user creating the dynamic Windows desktop needs to have a role with access to the
Windows desktop labels and the `dynamic_windows_desktop` resource.  In this example role the user can only
create and maintain dynamic Windows desktops labeled `env: test`.

```yaml
kind: role
metadata:
  name: example
spec:
  allow:
    windows_desktop_labels:
      env: test
    rules:

>     - resources:
>       - dynamic_windows_desktop
>       verbs: [list, create, read, update, delete]

Reviewer: ```suggestion
    - resources: [dynamic_windows_desktop]
      verbs: [list, create, read, update, delete]
```

I find this slightly easier to read

version: v7
```

To create a dynamic Windows desktop resource, run:

<Tabs>
    <TabItem scope={["oss", "enterprise"]} label="Self-Hosted">

        ```code
        # Log in to your cluster with tsh so you can use tctl from your local machine.
        # You can also run tctl on your Auth Service host without running "tsh login"
        # first.
        $ tsh login --proxy=teleport.example.com --user=myuser
        $ tctl create dynamic_windows_desktop.yaml
        ```

    </TabItem>
    <TabItem scope={["cloud"]} label="Teleport Enterprise (Cloud)">

        ```code
        # Log in to your Teleport cluster so you can use tctl remotely.
        $ tsh login --proxy=mytenant.teleport.sh --user=myuser
        $ tctl create dynamic_windows_desktop.yaml
        ```

    </TabItem>

</Tabs>

After the resource has been created, it will appear among the list of available

> Windows desktops (in `tsh desktops ls` or UI) as long as at least one Windows Desktop Service

Reviewer: ```suggestion
Windows desktops (in the web UI) as long as at least one Windows Desktop Service
```

There's no such thing as `tsh desktops ls` - we never built it because you can't connect to desktops with tsh.
